File: backend/schedule/article.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
@Author: reber
@Mail: reber0ask@qq.com
@Date: 2021-01-06 09:40:37
@LastEditTime : 2021-01-31 14:24:23
'''

import re
import time
import requests
from sqlalchemy import func

from sqlmodule import session_maker
from sqlmodule import to_dict
from sqlmodule import Article
from sqlmodule import Data

from libs.auth import get_username
from libs.request import req, ReqExceptin
from libs.common import logger_msg

from setting import rss_sqlite_uri
from setting import logger


class ArticleClass(object):
    def __init__(self):
        super(ArticleClass, self).__init__()

    def get_all_site_msg(self):
        """
        获取所有站点的信息
        """
        site_msg_list = list()
        try:
            with session_maker(rss_sqlite_uri) as db_session:
                results = db_session.query(Article).all()
                site_msg_list = to_dict(results)
        except Exception as e:
            logger.error(str(e))
        else:
            return site_msg_list

    def get_site_article_url(self, site_id):
        """
        获取一个博客的现有文章链接
        """
        site_article_list = list()
        article_url_list = list()
        try:
            with session_maker(rss_sqlite_uri) as db_session:
                results = db_session.query(Data).filter_by(ref_id=site_id).filter_by(category='article').all()
                site_article_list = to_dict(results)
        except Exception as e:
            logger.error(str(e))
        else:
            if site_article_list:
                article_url_list = [article_msg["url"] for article_msg in site_article_list]
        return article_url_list

    def get_site_new_article_msg(self, username, name, link, regex, article_url_list):
        """
        获取一个站点的新文章信息
        """
        error_msg = ""
        new_article_msg_list = list()
        try:
            resp = req.get(url=link)
            html = resp.content
        except ReqExceptin as error_msg:
            logger.error(error_msg)
            logger_msg(msg_type="system", username="schedule", action="article check: {}".format(name), data=str(error_msg))
            logger_msg(msg_type="user", username=username, action="{} 更新".format(name), data=str(error_msg))
        else:
            href_text_list = re.findall(regex, str(html, encoding='utf-8'), re.S|re.M)

            for href_text in href_text_list:
                a_href = href_text[0].strip()
                if not a_href.startswith("http"):
                    a_href = link.rstrip("/")+"/"+a_href.lstrip("/")
                a_text = href_text[1].strip()

                if a_href not in article_url_list:
                    # resp = requests.get(url=a_href, verify=False)
                    # html = str(resp.content, encoding='utf-8')
                    # html = "<iframe src={}></iframe>".format(a_href)
                    html = "<h2>双指向左滑动</h2>"
                    new_article_msg_list.insert(0, (a_text, a_href, html))

        return new_article_msg_list

    def check(self):
        all_site_msg = self.get_all_site_msg()
        for site_msg in all_site_msg:
            site_id = site_msg.get("id")
            user_id = site_msg.get("user_id")
            name = site_msg.get("name")
            link = site_msg.get("link")
            regex = site_msg.get("regex")
            username = get_username(user_id=user_id)

            # logger.info("Article check: {}".format(link))
            article_url_list = self.get_site_article_url(site_id)
            new_article_msg_list = self.get_site_new_article_msg(username, name, link, regex, article_url_list)

            article_data = list()
            for new_article_msg in new_article_msg_list:
                title, url, html = new_article_msg
                date = func.now()

                article = Data(ref_id=site_id, category="article", title=title, url=url, date=date, description=html, status="unread")
                article_data.append(article)

            if article_data:
                message_data_list = [x[0] for x in new_article_msg_list] # 获取 title 的 list 用于保存
                logger_msg(msg_type="system", username="schedule", action="article check: {}".format(name), data=str(message_data_list))
                logger_msg(msg_type="user", username=username, action="{} 已更新".format(name), data=str(message_data_list))
                try:
                    with session_maker(rss_sqlite_uri) as db_session:
                        db_session.add_all(article_data)
                except Exception as e:
                    logger.error(str(e))
                else:
                    logger.info("Article check: {} 获取 {} 个新资源".format(name, len(article_data)))

def article_check():
    logger.info("==> article check...")

    logger_msg(msg_type="system", action="start article check", username="schedule")
    _article = ArticleClass()
    _article.check()

File: server/schedule/article.go
/*
 * @Author: reber
 * @Mail: reber0ask@qq.com
 * @Date: 2022-01-04 21:12:25
 * @LastEditTime: 2024-12-04 13:24:40
 */
package schedule

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/reber0/goutils"
	"github.com/reber0/rss-sub/server/global"
	"github.com/reber0/rss-sub/server/mydb"
)

func checkArticle() {
	global.Log.Info("checkArticle...")

	articleAllSiteMsg := getArticleAllSiteMsg()
	for _, siteMsg := range articleAllSiteMsg {
		site_id := siteMsg.ID
		user_id := siteMsg.UID
		name := siteMsg.Name
		link := siteMsg.Link
		regex := siteMsg.Regex

		var username string
		result := global.Db.Model(&mydb.User{}).Select("uname").Where("uid=?", user_id).First(&username)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
		}

		global.Log.Info(fmt.Sprintf("schedule article check ==> %s", name))
		articleURLSlice := getArticleURL(site_id)
		newArticleMsgList, err := getNewArticleMsg(link, regex, articleURLSlice)
		if err != nil {
			loggerMsg(username, name+" 更新", err.Error())
		} else {
			if len(newArticleMsgList) > 0 {
				// 保存新的文章信息
				var titleSlice []string
				for _, newArticleMsg := range newArticleMsgList {
					title := newArticleMsg[0]
					url := newArticleMsg[1]
					data := mydb.Data{RefId: site_id, Category: "article", Title: title, URL: url, Status: "unread"}
					global.Db.Create(&data)

					titleSlice = append(titleSlice, title)
				}
				loggerMsg(username, name+" 更新", strings.Join(titleSlice, ","))
			}
		}
	}
}

// 获取所有站点的信息
func getArticleAllSiteMsg() []mydb.Article {
	var siteMsgSlice []mydb.Article
	result := global.Db.Model(&mydb.Article{}).Find(&siteMsgSlice)
	if result.Error != nil {
		global.Log.Error(result.Error.Error())
	}
	return siteMsgSlice
}

// 获取一个博客已获取的文章链接
func getArticleURL(articleID int) []string {
	var articleURLSlice []string

	var dataSlice []mydb.Data
	result := global.Db.Model(&mydb.Data{}).Where("category='article' and ref_id = ?", articleID).Find(&dataSlice)
	if result.Error != nil {
		global.Log.Error(result.Error.Error())
	} else {
		for _, data := range dataSlice {
			articleURLSlice = append(articleURLSlice, data.URL)
		}
	}

	return articleURLSlice
}

// 获取一个博客更新的文章信息
func getNewArticleMsg(link, regex string, articleURLSlice []string) ([][]string, error) {
	var newArticleMsgList [][]string

	resp, err := global.Client.R().Get(link)
	if err != nil {
		global.Log.Error(err.Error())
		return newArticleMsgList, err
	} else {
		html := goutils.EncodeToUTF8(resp)
		baseURL := goutils.NewURL(link).BaseURL()

		reg := regexp.MustCompile(`(?sm)` + regex)
		href_text_list := reg.FindAllStringSubmatch(html, -1)
		for _, href_text := range href_text_list {
			a_href := strings.TrimSpace(href_text[1])
			if !strings.HasPrefix(a_href, "http") {
				a_href = baseURL + strings.TrimLeft(a_href, "/")
			}
			a_text := strings.TrimSpace(href_text[2])

			// 暂存新文章
			if !goutils.IsInCol(a_href, articleURLSlice) {
				newArticleMsgList = append(newArticleMsgList, []string{a_text, a_href})
			}
		}
		newArticleMsgList = goutils.SliceListReverse(newArticleMsgList)
	}

	return newArticleMsgList, nil
}
